File: apps/api/app/schemas/holding.py
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field


class HoldingCreate(BaseModel):
    ticker: str = Field(..., min_length=1, max_length=10, pattern=r"^[A-Za-z.]+$")
    shares: float = Field(..., gt=0)
    purchased_at: date | None = None
    cost_basis: float | None = Field(None, ge=0)


class HoldingUpdate(BaseModel):
    shares: float = Field(..., gt=0)
    purchased_at: date | None = None
    cost_basis: float | None = Field(None, ge=0)


class HoldingRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    portfolio_id: int
    ticker: str
    shares: float
    purchased_at: date | None = None
    cost_basis: float | None = None
    last_price: float | None = None
    previous_close: float | None = None
    sector: str | None = None
    beta: float | None = None
    dividend_yield: float | None = None
    next_earnings_date: str | None = None
    latest_earnings_call: datetime | None = None
    earnings_call_summary: str | None = None
    created_at: datetime
    updated_at: datetime

    @property
    def value(self) -> float:
        return (self.last_price or 0.0) * self.shares

File: apps/web/src/types/holding.ts
export interface HoldingRead {
  id: number;
  portfolio_id: number;
  ticker: string;
  shares: number;
  purchased_at: string | null;
  cost_basis: number | null;
  last_price: number | null;
  previous_close: number | null;
  sector: string | null;
  beta: number | null;
  dividend_yield: number | null;
  next_earnings_date: string | null;
  latest_earnings_call: string | null;
  earnings_call_summary: string | null;
  created_at: string;
  updated_at: string;
}

export interface HoldingCreate {
  ticker: string;
  shares: number;
  purchased_at?: string | null;
  cost_basis?: number | null;
}

export interface HoldingUpdate {
  shares: number;
  purchased_at?: string | null;
  cost_basis?: number | null;
}

export function holdingValue(h: HoldingRead): number {
  return (h.last_price ?? 0) * h.shares;
}

export function holdingGainLoss(h: HoldingRead): number | null {
  if (h.cost_basis == null || h.last_price == null) return null;
  return (h.last_price - h.cost_basis) * h.shares;
}

export function holdingGainLossPct(h: HoldingRead): number | null {
  if (h.cost_basis == null || h.cost_basis === 0 || h.last_price == null) return null;
  return ((h.last_price - h.cost_basis) / h.cost_basis) * 100;
}
